File: cn/java-springmvc/src/main/java/com/example/controller/ReviewController.java
package com.example.controller;

import org.springframework.web.bind.annotation.*;
import com.example.model.Review;
import com.example.dto.Result;
import com.example.dto.PageResult;
import jakarta.validation.Valid;
import java.util.List;

@RestController
@RequestMapping("/api/reviews")
public class ReviewController {
    
    @GetMapping
    public Result<PageResult<Review>> listReviews(
            @RequestParam(required = false) Long userId,
            @RequestParam(required = false) Long productId,
            @RequestParam(required = false) Integer rating,
            @RequestParam(required = false) Integer status,
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @PostMapping
    public Result<Review> createReview(@Valid @RequestBody CreateReviewReq req) {
        return Result.success(new Review());
    }
    
    @GetMapping("/{id}")
    public Result<Review> getReview(@PathVariable Long id) {
        return Result.success(new Review());
    }
    
    @PutMapping("/{id}")
    public Result<Review> updateReview(@PathVariable Long id, @Valid @RequestBody UpdateReviewReq req) {
        return Result.success(new Review());
    }
    
    @DeleteMapping("/{id}")
    public Result<Void> deleteReview(@PathVariable Long id) {
        return Result.success(null);
    }
    
    @PostMapping("/{id}/like")
    public Result<Review> likeReview(@PathVariable Long id) {
        return Result.success(new Review());
    }
    
    @DeleteMapping("/{id}/like")
    public Result<Review> unlikeReview(@PathVariable Long id) {
        return Result.success(new Review());
    }
    
    @PostMapping("/{id}/reply")
    public Result<ReviewReplyVO> replyToReview(@PathVariable Long id, @Valid @RequestBody CreateReplyReq req) {
        return Result.success(new ReviewReplyVO());
    }
    
    @GetMapping("/{id}/replies")
    public Result<List<ReviewReplyVO>> getReplies(@PathVariable Long id) {
        return Result.success(List.of());
    }
    
    @GetMapping("/product/{productId}")
    public Result<PageResult<Review>> getProductReviews(
            @PathVariable Long productId,
            @RequestParam(required = false) Integer rating,
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @GetMapping("/user/{userId}")
    public Result<PageResult<Review>> getUserReviews(
            @PathVariable Long userId,
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @PutMapping("/{id}/status")
    public Result<Review> updateStatus(@PathVariable Long id, @RequestParam Integer status) {
        return Result.success(new Review());
    }
    
    @GetMapping("/product/{productId}/statistics")
    public Result<ReviewStatisticsVO> getProductStatistics(@PathVariable Long productId) {
        return Result.success(new ReviewStatisticsVO());
    }
    
    @GetMapping("/product/{productId}/rating-distribution")
    public Result<List<RatingDistributionVO>> getRatingDistribution(@PathVariable Long productId) {
        return Result.success(List.of());
    }
    
    @GetMapping("/recent")
    public Result<List<Review>> getRecentReviews(@RequestParam(defaultValue = "10") Integer limit) {
        return Result.success(List.of());
    }
    
    @GetMapping("/helpful")
    public Result<List<Review>> getHelpfulReviews(
            @RequestParam(defaultValue = "10") Integer limit) {
        return Result.success(List.of());
    }
    
    @PostMapping("/batch-approve")
    public Result<Void> batchApprove(@RequestBody List<Long> ids) {
        return Result.success(null);
    }
    
    @PostMapping("/batch-reject")
    public Result<Void> batchReject(@RequestBody List<Long> ids, @RequestParam String reason) {
        return Result.success(null);
    }
    
    @GetMapping("/pending")
    public Result<PageResult<Review>> getPendingReviews(
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @GetMapping("/export")
    public Result<String> exportReviews(
            @RequestParam(required = false) Long productId,
            @RequestParam(required = false) String startDate,
            @RequestParam(required = false) String endDate) {
        return Result.success("export_file_url");
    }
    
    @GetMapping("/{id}/images")
    public Result<List<String>> getReviewImages(@PathVariable Long id) {
        return Result.success(List.of());
    }
    
    @DeleteMapping("/{id}/images")
    public Result<Void> deleteReviewImage(@PathVariable Long id, @RequestParam String imageUrl) {
        return Result.success(null);
    }
}

class CreateReviewReq {
    private Long userId;
    private Long productId;
    private Long orderId;
    private Integer rating;
    private String content;
    private String images;
    
    public Long getUserId() { return userId; }
    public void setUserId(Long userId) { this.userId = userId; }
    public Long getProductId() { return productId; }
    public void setProductId(Long productId) { this.productId = productId; }
    public Long getOrderId() { return orderId; }
    public void setOrderId(Long orderId) { this.orderId = orderId; }
    public Integer getRating() { return rating; }
    public void setRating(Integer rating) { this.rating = rating; }
    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }
    public String getImages() { return images; }
    public void setImages(String images) { this.images = images; }
}

class UpdateReviewReq {
    private Integer rating;
    private String content;
    private String images;
    
    public Integer getRating() { return rating; }
    public void setRating(Integer rating) { this.rating = rating; }
    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }
    public String getImages() { return images; }
    public void setImages(String images) { this.images = images; }
}

class CreateReplyReq {
    private Long userId;
    private String content;
    
    public Long getUserId() { return userId; }
    public void setUserId(Long userId) { this.userId = userId; }
    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }
}

class ReviewReplyVO {
    private Long id;
    private Long reviewId;
    private Long userId;
    private String username;
    private String content;
    private String createdAt;
    
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public Long getReviewId() { return reviewId; }
    public void setReviewId(Long reviewId) { this.reviewId = reviewId; }
    public Long getUserId() { return userId; }
    public void setUserId(Long userId) { this.userId = userId; }
    public String getUsername() { return username; }
    public void setUsername(String username) { this.username = username; }
    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }
    public String getCreatedAt() { return createdAt; }
    public void setCreatedAt(String createdAt) { this.createdAt = createdAt; }
}

class ReviewStatisticsVO {
    private Long productId;
    private Long totalReviews;
    private Double averageRating;
    private Integer fiveStarCount;
    private Integer fourStarCount;
    private Integer threeStarCount;
    private Integer twoStarCount;
    private Integer oneStarCount;
    
    public Long getProductId() { return productId; }
    public void setProductId(Long productId) { this.productId = productId; }
    public Long getTotalReviews() { return totalReviews; }
    public void setTotalReviews(Long totalReviews) { this.totalReviews = totalReviews; }
    public Double getAverageRating() { return averageRating; }
    public void setAverageRating(Double averageRating) { this.averageRating = averageRating; }
    public Integer getFiveStarCount() { return fiveStarCount; }
    public void setFiveStarCount(Integer fiveStarCount) { this.fiveStarCount = fiveStarCount; }
    public Integer getFourStarCount() { return fourStarCount; }
    public void setFourStarCount(Integer fourStarCount) { this.fourStarCount = fourStarCount; }
    public Integer getThreeStarCount() { return threeStarCount; }
    public void setThreeStarCount(Integer threeStarCount) { this.threeStarCount = threeStarCount; }
    public Integer getTwoStarCount() { return twoStarCount; }
    public void setTwoStarCount(Integer twoStarCount) { this.twoStarCount = twoStarCount; }
    public Integer getOneStarCount() { return oneStarCount; }
    public void setOneStarCount(Integer oneStarCount) { this.oneStarCount = oneStarCount; }
}

class RatingDistributionVO {
    private Integer rating;
    private Integer count;
    private Double percentage;
    
    public Integer getRating() { return rating; }
    public void setRating(Integer rating) { this.rating = rating; }
    public Integer getCount() { return count; }
    public void setCount(Integer count) { this.count = count; }
    public Double getPercentage() { return percentage; }
    public void setPercentage(Double percentage) { this.percentage = percentage; }
}

File: cn/kotlin-springmvc/src/main/kotlin/com/example/controller/ReviewController.kt
package com.example.controller

import com.example.dto.PageResult
import com.example.dto.Result
import com.example.model.Review
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/reviews")
class ReviewController {
    
    @GetMapping
    fun listReviews(
        @RequestParam(required = false) productId: Long?,
        @RequestParam(required = false) userId: Long?,
        @RequestParam(required = false) rating: Int?,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Review>> {
        val result = PageResult<Review>()
        return Result.success(result)
    }
    
    @PostMapping
    fun createReview(@RequestBody req: CreateReviewReq): Result<Review> {
        val review = Review()
        return Result.success(review)
    }
    
    @GetMapping("/{id}")
    fun getReview(@PathVariable id: Long): Result<Review> {
        val review = Review()
        return Result.success(review)
    }
    
    @PutMapping("/{id}")
    fun updateReview(@PathVariable id: Long, @RequestBody req: UpdateReviewReq): Result<Review> {
        val review = Review()
        return Result.success(review)
    }
    
    @DeleteMapping("/{id}")
    fun deleteReview(@PathVariable id: Long): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/product/{productId}")
    fun getProductReviews(
        @PathVariable productId: Long,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Review>> {
        val result = PageResult<Review>()
        return Result.success(result)
    }
    
    @GetMapping("/user/{userId}")
    fun getUserReviews(
        @PathVariable userId: Long,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Review>> {
        val result = PageResult<Review>()
        return Result.success(result)
    }
    
    @GetMapping("/product/{productId}/statistics")
    fun getProductStatistics(@PathVariable productId: Long): Result<ReviewStatisticsVO> {
        val statistics = ReviewStatisticsVO()
        return Result.success(statistics)
    }
    
    @PutMapping("/{id}/helpful")
    fun markHelpful(@PathVariable id: Long): Result<Review> {
        val review = Review()
        return Result.success(review)
    }
    
    @PutMapping("/{id}/status")
    fun updateStatus(@PathVariable id: Long, @RequestParam status: Int): Result<Review> {
        val review = Review()
        return Result.success(review)
    }
    
    @PostMapping("/{id}/reply")
    fun addReply(@PathVariable id: Long, @RequestBody req: CreateReplyReq): Result<ReviewReplyVO> {
        val reply = ReviewReplyVO()
        return Result.success(reply)
    }
    
    @GetMapping("/{id}/replies")
    fun getReplies(@PathVariable id: Long): Result<List<ReviewReplyVO>> {
        val replies = listOf<ReviewReplyVO>()
        return Result.success(replies)
    }
    
    @GetMapping("/product/{productId}/rating-distribution")
    fun getRatingDistribution(@PathVariable productId: Long): Result<List<RatingDistributionVO>> {
        val distribution = listOf<RatingDistributionVO>()
        return Result.success(distribution)
    }
    
    @GetMapping("/recent")
    fun getRecentReviews(@RequestParam(defaultValue = "10") limit: Int): Result<List<Review>> {
        val reviews = listOf<Review>()
        return Result.success(reviews)
    }
    
    @GetMapping("/top-rated")
    fun getTopRatedReviews(@RequestParam(defaultValue = "10") limit: Int): Result<List<Review>> {
        val reviews = listOf<Review>()
        return Result.success(reviews)
    }
    
    @GetMapping("/pending")
    fun getPendingReviews(
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Review>> {
        val result = PageResult<Review>()
        return Result.success(result)
    }
    
    @PostMapping("/batch-approve")
    fun batchApprove(@RequestBody ids: List<Long>): Result<Void> {
        return Result.success(null)
    }
    
    @PostMapping("/batch-reject")
    fun batchReject(@RequestBody ids: List<Long>): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/{id}/images")
    fun getReviewImages(@PathVariable id: Long): Result<List<String>> {
        val images = listOf<String>()
        return Result.success(images)
    }
    
    @PostMapping("/{id}/images")
    fun addImages(@PathVariable id: Long, @RequestBody imageUrls: List<String>): Result<Review> {
        val review = Review()
        return Result.success(review)
    }
    
    @DeleteMapping("/{id}/images")
    fun deleteImage(@PathVariable id: Long, @RequestParam imageUrl: String): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/export")
    fun exportReviews(
        @RequestParam(required = false) productId: Long?,
        @RequestParam(required = false) startDate: String?,
        @RequestParam(required = false) endDate: String?
    ): Result<String> {
        val fileUrl = ""
        return Result.success(fileUrl)
    }
    
    @GetMapping("/product/{productId}/can-review")
    fun canUserReview(
        @PathVariable productId: Long,
        @RequestParam userId: Long
    ): Result<Boolean> {
        val canReview = true
        return Result.success(canReview)
    }
}

data class CreateReviewReq(
    val productId: Long?,
    val userId: Long?,
    val orderId: Long?,
    val rating: Int?,
    val title: String?,
    val content: String?,
    val images: List<String>?
)

data class UpdateReviewReq(
    val rating: Int?,
    val title: String?,
    val content: String?
)

data class ReviewStatisticsVO(
    val productId: Long? = null,
    val averageRating: Double? = null,
    val totalReviews: Int? = null,
    val fiveStarCount: Int? = null,
    val fourStarCount: Int? = null,
    val threeStarCount: Int? = null,
    val twoStarCount: Int? = null,
    val oneStarCount: Int? = null
)

data class ReviewReplyVO(
    val id: Long? = null,
    val reviewId: Long? = null,
    val content: String? = null,
    val replierName: String? = null,
    val createdAt: String? = null
)

data class CreateReplyReq(
    val content: String?
)

data class RatingDistributionVO(
    val rating: Int? = null,
    val count: Int? = null,
    val percentage: Double? = null
)
